src/shuffler-codereview.c

```diff
@@ -0,0 +1,1228 @@
+    ShuffleStarters(&masterSeed);
+
+    // give the player some items to start with.
+    AddInitialBagItems(&masterSeed);
+
+    // generate the first set of rooms.
+    GenerateNextWave();
+}
+
+bool8 IsInvalidWarp(struct WarpData* warp) {
+    return warp->mapGroup == -1 && warp->mapNum == -1;
+}
+
+bool8 PokemonIsValidLevel(const struct PokemonTemplate* pokemon, u8 level) {
+    if (pokemon->minLevel != 0 && pokemon->minLevel > level) {
+        return FALSE;
+    }
+    if (pokemon->maxLevel != 0 && pokemon->maxLevel < level) {
+        return FALSE;
+    }
+    return TRUE;
+};
+
+void MirrorMapData() {
+    u16 currentRoom = GetRoomId(gSaveBlock1Ptr->location.mapNum,gSaveBlock1Ptr->location.mapGroup);
+    if (CurrentAdjustedRoom == currentRoom) {
+        return;
+    }
+
+    CurrentAdjustedRoom = currentRoom;
+    AdjustedObjectsDirty = 0;
+    RedirectedTrainerFlags = 0;
+
+    MYLOG("distance: %u, type: %u", distance, GetCurrentAreaType());
+
+    AdjustedMapEvents = *gMapHeader.events;
+    
+    u8 num_objs = AdjustedMapEvents.objectEventCount;
+    if (num_objs > MAX_OBJECTS) {
+        MYLOG("num_objs > MAX_OBJECTS, shit is broken: %d", num_objs);
+    } else {
+        size_t objs_bytes = num_objs * sizeof(struct ObjectEventTemplate);
+        memcpy(&AdjustedTemplates, AdjustedMapEvents.objectEvents, objs_bytes);
+        AdjustedMapEvents.objectEvents = &AdjustedTemplates[0];
+    }
+    gMapHeader.events = &AdjustedMapEvents;
+}
+
+void DeclareTrainer(u8 objNum, u8 trainerType, u8 partnerObjNum) {
+    MirrorMapData();
+    AdjustedObjectsDirty |= (1 << objNum);
+
+    const struct TrainerTemplate *tt;
+    if (trainerType == TRAINER_TYPE_GYM_MEMBER) {
+        tt = GetGymMember(NextGymType, NextGymMinorType);
+    } else if(trainerType == TRAINER_TYPE_GYM_LEADER) {
+        tt = GetGymLeader(NextGymType, NextGymMinorType);
+    } else {
+        int r = tinymt32_generate_uint32(&masterSeed) % GetTrainerGroupCount(trainerType);
+        tt = GetTrainerGroup(trainerType)[r];
+    }
+
+    struct ShuffledTrainerInfo* trainerInfo = &AdjustedObjects[objNum].t;
+    struct Trainer* trainer = &trainerInfo->trainer;
+
+    trainer->aiFlags = AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_TRY_TO_FAINT | AI_FLAG_CHECK_VIABILITY;
+    trainer->doubleBattle = FALSE;
+    trainer->encounterMusic_gender = tt->encounterMusic_gender;
+    trainer->trainerClass = tt->trainerClass;
+    trainer->trainerPic = tt->trainerPic;
+    for (int i = 0; i < 4; i++) {
+        trainer->items[i] = ITEM_NONE;
+    }
+
+    trainer->partyFlags = F_TRAINER_PARTY_CUSTOM_MOVESET | F_TRAINER_PARTY_HELD_ITEM;
+    u8 level = GetScaledLevel(distance);
+    int partyIndicies[6];
+    switch (tt->teamBuildingStrategy) {
+    case TB_STRAT_STATIC:
+        trainer->partySize = tt->pool1Size;
+        for (int i = 0; i < tt->pool1Size; i++) {
+            partyIndicies[i] = i;
+        }
+        break;
+    case TB_STRAT_COMMON: {
```

Comment:
throughout, your case statements are MASSIVE. Might want to break them into smaller chunks with function calls. (switch statements have weird and confusing scoping rules. You're actually bulletproofing your code by doing this.)